api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -742,6 +743,50 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
     return ResponseEntity.ok(response);
   }
 
+  @Override
+  public ResponseEntity<FileDetail> renameNotebook(String workspace, String workspaceName, NotebookRename rename) {
+    String fromBucket = fireCloudService.getWorkspace(workspace, workspaceName)
+      .getWorkspace()
+      .getBucketName();
+    String origPath = NOTEBOOKS_WORKSPACE_DIRECTORY+ "/" + rename.getName();
+    String newName = rename.getNewName();
+    if (!newName.matches("^.+\\.ipynb")) {
+      newName = newName + ".ipynb";
+    }
+    String newPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + newName;
+    BlobId blobId = BlobId.of(fromBucket, origPath);
+    cloudStorageService.copyBlob(blobId, BlobId.of(fromBucket, newPath));
+    cloudStorageService.deleteBlob(blobId);
+    FileDetail fileDetail = new FileDetail();
+    fileDetail.setName(newName);
+    fileDetail.setPath("gs://" + fromBucket + "/" + NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + newName);
+    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+    fileDetail.setLastModifiedTime(now.getTime());
+    return ResponseEntity.ok(fileDetail);
+  }
+
+  @Override
+  public ResponseEntity<EmptyResponse> cloneNotebook(String workspace, String workspaceName, String notebookName) {
+    String fromBucket = fireCloudService.getWorkspace(workspace, workspaceName)
```

Comment:
nit: I would probably call this just `bucket` here and above. I used `from` and `to` prefixes in similar methods just because the source and destination were different.